massa-execution-worker/src/execution.rs

```diff
@@ -404,12 +404,53 @@ impl ExecutionState {
     fn process_denunciation(
         &self,
         denunciation: &Denunciation,
+        block_slot: &Slot,
         block_credits: &mut Amount,
     ) -> Result<(), ExecutionError> {
         let addr_denounced = Address::from_public_key(denunciation.get_public_key());
 
         // acquire write access to the context
         let mut context = context_guard!(self);
+
+        // ignore denunciation if not valid
+        if !denunciation.is_valid() {
+            return Err(ExecutionError::IncludeDenunciationError(
+                "denunciation is not valid".to_string(),
+            ));
+        }
+
+        // ignore denunciation if too old or expired
+        let de_slot = denunciation.get_slot();
+
+        if de_slot.period
+            < self
+                .final_cursor
+                .period
+                .saturating_sub(self.config.denunciation_expire_periods)
+        {
+            // too old - cannot be denounced anymore
+            return Err(ExecutionError::IncludeDenunciationError(
+                "denunciation is too old".to_string(),
+            ));
+        }
+
+        if de_slot.period.saturating_sub(block_slot.period)
+            > self.config.denunciation_expire_periods
+        {
+            // too much in the future - ignored
+            return Err(ExecutionError::IncludeDenunciationError(
+                "denunciation is too much in the future".to_string(),
+            ));
+        }
```

Comment:
This is correct for the pool but incorrect here.

Here we should not look at finality at all and we should check:
* if de_slot > block_slot => IGNORE:  a block at slot S cannot denounce something that is at a slot > S
  * note we should ensure that de_slot == block_slot is OK, for example if the block producer wants to denounce someone who multi-endorsed for the block's slot
* if de_slot.period < block_slot.period - denuniciation_periods => this block cannot include a denunciation that denounces something that is more than denuniciation_periods before the block

Reply:
Fxied